linux/ibus-keyman/src/bcp47util.c

```diff
@@ -0,0 +1,99 @@
+#include <gio/gio.h>
+#include <unicode/locid.h>
+#include <unicode/uloc.h>
+
+/// Minimize the BCP-47 `tag` so that unnecessary parts get ommitted.
+/// The result gets stored in `minimzedTag`.
+///
+/// @param tag           The tag to process
+/// @param minimizedTag  Caller-provided character array for the
+///                      resulting minimized tag.
+/// @param tagCapacity   Array size of `minimizedTag`
+/// @return The length of the minimized tag, or -1 in error case
+int bcp47_minimize(const char* tag, char* minimizedTag, int tagCapacity) {
+  UErrorCode status = U_ZERO_ERROR;
+  if (!tag || strlen(tag) == 0) {
+    strncpy(minimizedTag, "", tagCapacity);
+    return -1;
+  }
+
+  // special treatment for `und-Latn` which is used by sil_ipa keyboard
+  if (strcmp(tag, "und-Latn") == 0) {
+    strncpy(minimizedTag, "und-Latn", tagCapacity);
+    return strlen(minimizedTag);
+  }
+
+  int capacity = 255;
+  char workingTag[capacity];
+
+  // special treatment for tags that start with `und`: replace `und` with `en`.
+  // ICU 70 doesn't properly treat `und`.
+  int isUnd = strncmp(tag, "und", 3) == 0;
+  if (isUnd) {
+    strcpy(workingTag, "en");
+    strncat(&workingTag[2], &tag[3], capacity - 3);
+  } else {
+    strncpy(workingTag, tag, capacity - 1);
+  }
+  workingTag[capacity - 1] = 0;
+
+  char localeId[capacity];
+  uloc_forLanguageTag(workingTag, localeId, capacity, NULL, &status);
```

Comment:
uloc_forLanguageTag(`und`) will be interesting, maybe not what's expected - may be root